### PR TITLE
new block format pt2 -- remove specials

### PR DESCRIPTION
**Pull request by djrtwo**
part 2 in a series of PRs addressing #246.

* removes `block.body.specials`
* move all specials to specific block objects contained in `block.body`
* Create section "Block objects" in per-slot processing for all these `body` object types

### CODE REVIEW

File: specs/core/0_beacon-chain.md [hunk 1]
```diff
@@ -200,14 +200,15 @@ Unless otherwise indicated, code appearing in `this style` is to be interpreted
 | `EXITED_WITHOUT_PENALTY` | `3` |
 | `EXITED_WITH_PENALTY` | `4` |
 
-### Special record types
+### Max transactions
```
Comment (vbuterin) [inline specs/core/0_beacon-chain.md]: Do we really want to call them "transactions"? I would be inclined to save that term for user-defined activity, and call them something more neutral, eg. "records".

File: specs/core/0_beacon-chain.md [hunk 2]
```diff
@@ -317,27 +321,106 @@ Unless otherwise indicated, code appearing in `this style` is to be interpreted
...
+        'deposit_parameters': DepositParametersRecord,
+        # Value in Gwei
+        'value': 'uint64',
+        # Timestamp from deposit contract
+        'timestamp': 'uint64',
+    },
+}
+```
+
+#### `DepositParametersRecord`
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]:
```suggestion
##### `DepositParametersRecord`
```

File: specs/core/0_beacon-chain.md [hunk 3]
```diff
@@ -317,27 +285,130 @@ Unless otherwise indicated, code appearing in `this style` is to be interpreted
...
+}
+```
+
+#### `BeaconBlock`
+
+```python
+{
+    'header': BeaconBlockHeader,
+    'body': BeaconBlockBody,
+}
```
Comment (vbuterin) [inline specs/core/0_beacon-chain.md]: Not sure I like this; I'd feel better with `body` being an element of the header.

Remember that at the SSZ tree hashing level, that would just mean that the body gets represented by a nice clean 32 byte hash.

Comment (djrtwo) [reply specs/core/0_beacon-chain.md]: Are you suggesting the following? Just doing away with `BeaconBlockHeader`? I like that.

```
#### BeaconBlock

{
    'slot': 'uint64',
    # Skip list of ancestor beacon block hashes
    # i'th item is the most recent ancestor whose slot is a multiple of 2**i for i = 0, ..., 31
    'ancestor_hashes': ['hash32'],
    'state_root': 'hash32',
    'randao_reveal': 'hash32',
    'candidate_pow_receipt_root': 'hash32',
    'signature': ['uint384'],
    'body': BeaconBlockBody,
}

File: specs/core/0_beacon-chain.md [hunk 4]
```diff
@@ -1262,61 +1265,63 @@ For each `attestation` in `block.body.attestations`:
...
-#### `VOLUNTARY_EXIT`
+For each `proposer_slashing` in `block.body.proposer_slashings`:
 
-* Let `validator = state.validator_registry[validator_index]`.
-* Verify that `BLSVerify(pubkey=validator.pubkey, msg=ZERO_HASH, sig=signature, domain=get_domain(state.fork_data, slot, DOMAIN_EXIT))`.
-* Verify that `validator.status == ACTIVE`.
-* Verify that `state.slot >= slot`.
-* Verify that `state.slot >= validator.latest_status_change_slot + SHARD_PERSISTENT_COMMITTEE_CHANGE_PERIOD`.
-* Run `exit_validator(validator_index, state, penalize=False, current_slot=state.slot)`.
+* Let `proposer = state.validator_registry[proposer_slashing.proposer_index]`
```
Comment (JustinDrake) [inline specs/core/0_beacon-chain.md]: Missing period at the end

File: specs/core/0_beacon-chain.md [hunk 5]
```diff
@@ -226,54 +231,58 @@ Unless otherwise indicated, code appearing in `this style` is to be interpreted
...
+    # Second proposal signature
+    'proposal_signature_2': '[uint384]',
 }
 ```
 
-### Beacon chain blocks
+#### Casper slashings
 
-#### `BeaconBlock`
+##### `CasperSlashingRecord`
```
Comment (JustinDrake) [inline specs/core/0_beacon-chain.md]: I would agree with Vitalik that we should keep `Record` for objects stored in the state. What about `CasperSlashingObject` or `CasperSlashingSpecial`?

File: specs/core/0_beacon-chain.md [hunk 6]
```diff
@@ -1262,61 +1265,63 @@ For each `attestation` in `block.body.attestations`:
...
-* Verify that `BLSVerify(pubkey=validator.pubkey, msg=ZERO_HASH, sig=signature, domain=get_domain(state.fork_data, slot, DOMAIN_EXIT))`.
-* Verify that `validator.status == ACTIVE`.
-* Verify that `state.slot >= slot`.
-* Verify that `state.slot >= validator.latest_status_change_slot + SHARD_PERSISTENT_COMMITTEE_CHANGE_PERIOD`.
-* Run `exit_validator(validator_index, state, penalize=False, current_slot=state.slot)`.
+* Let `proposer = state.validator_registry[proposer_slashing.proposer_index]`
+* Verify that `BLSVerify(pubkey=proposer.pubkey, msg=hash(proposer_slashing.proposal_data_1), sig=proposer_slashing.proposal_signature_1, domain=get_domain(state.fork_data, proposer_slashing.proposal_data_1.slot, DOMAIN_PROPOSAL))`.
+* Verify that `BLSVerify(pubkey=proposer.pubkey, msg=hash(proposer_slashing.proposal_data_2), sig=proposer_slashing.proposal_signature_2, domain=get_domain(state.fork_data, proposer_slashing.proposal_data_2.slot, DOMAIN_PROPOSAL))`.
+* Verify that `proposer_slashing.proposal_data_1 != proposer_slashing.proposal_data_2`.
+* Verify that `proposer_slashing.proposal_data_1.slot == proposer_slashing.proposal_data_2.slot`.
```
Comment (JustinDrake) [inline specs/core/0_beacon-chain.md]: I think we should also check that `proposer_slashing.proposal_data_1.shard == proposer_slashing.proposal_data_2.shard`.

File: specs/core/0_beacon-chain.md [hunk 7]
```diff
@@ -1262,61 +1265,63 @@ For each `attestation` in `block.body.attestations`:
...
-* Let `validator = state.validator_registry[validator_index]`.
-* Verify that `BLSVerify(pubkey=validator.pubkey, msg=ZERO_HASH, sig=signature, domain=get_domain(state.fork_data, slot, DOMAIN_EXIT))`.
-* Verify that `validator.status == ACTIVE`.
-* Verify that `state.slot >= slot`.
-* Verify that `state.slot >= validator.latest_status_change_slot + SHARD_PERSISTENT_COMMITTEE_CHANGE_PERIOD`.
-* Run `exit_validator(validator_index, state, penalize=False, current_slot=state.slot)`.
+* Let `proposer = state.validator_registry[proposer_slashing.proposer_index]`
+* Verify that `BLSVerify(pubkey=proposer.pubkey, msg=hash(proposer_slashing.proposal_data_1), sig=proposer_slashing.proposal_signature_1, domain=get_domain(state.fork_data, proposer_slashing.proposal_data_1.slot, DOMAIN_PROPOSAL))`.
+* Verify that `BLSVerify(pubkey=proposer.pubkey, msg=hash(proposer_slashing.proposal_data_2), sig=proposer_slashing.proposal_signature_2, domain=get_domain(state.fork_data, proposer_slashing.proposal_data_2.slot, DOMAIN_PROPOSAL))`.
+* Verify that `proposer_slashing.proposal_data_1 != proposer_slashing.proposal_data_2`.
```
Comment (JustinDrake) [inline specs/core/0_beacon-chain.md]: Suggest making this more explicit: `proposer_slashing.proposal_data_1.block_hash != proposer_slashing. proposal_data_2.block_hash`.

File: specs/core/0_beacon-chain.md [hunk 8]
```diff
@@ -1262,61 +1265,63 @@ For each `attestation` in `block.body.attestations`:
 * If `block.candidate_pow_receipt_root` is `x.candidate_pow_receipt_root` for some `x` in `state.candidate_pow_receipt_roots`, set `x.votes += 1`.
 * Otherwise, append to `state.candidate_pow_receipt_roots` a new `CandidatePoWReceiptRootRecord(candidate_pow_receipt_root=block.candidate_pow_receipt_root, votes=1)`.
 
-### Special objects
-
-* Verify that the quantity of each type of object in `block.body.specials` is less than or equal to its maximum (see table at the top).
-* Verify that objects are sorted in order of `kind`. That is, `block.body.specials[i+1].kind >= block.body.specials[i].kind` for `0 <= i < len(block.body.specials-1)`.
+### Operations
```
Comment (JustinDrake) [inline specs/core/0_beacon-chain.md]: "Operations" feels a bit awkward. What about "beacon transaction"?

Comment (djrtwo) [reply specs/core/0_beacon-chain.md]: I'm not totally against `Beacon transaction`, but @vbuterin wanted to reserve that term exclusively for user level activity.

I think operations is better here than "special" or "objects".

The following seems natural to me
- Verify the block header
- Verify and perform block operations
- Transit epoch if necessary

File: specs/core/0_beacon-chain.md (hunk 5 above)
Comment (djrtwo) [reply specs/core/0_beacon-chain.md]: In general object is implicit on all objects... but due to the conflicting debate on naming these things, I'll go with object for now.

Comment (djrtwo) [reply specs/core/0_beacon-chain.md]: I actually just removed the suffix from the objects. I think it is reasonable.

**Comment by djrtwo** [comment]
* stuck with "Operations" to describe handling the `block.body`
* removed `Record` from the suffix of these data structures
* addressed other PR comments
* handled merge conflict from master

Let me know thoughts/further review

**Comment by djrtwo** [comment]
cc: @JustinDrake @hwwhww for another look

**Comment by hwwhww** [comment]
Highlight @JustinDrake 's changelog:

### Casper cleanups and bug fixes

* (cleanup) `SpecialAttestationData` => `CasperVotes`
* (cleanup) `vote_1`, `vote_2` => `votes_1`, `votes_2`
* (bug fix) Place a bound on the number of votes with `MAX_CASPER_VOTES`
* (bug fix) Supply `state` when calling `verify_casper_votes`

File: specs/core/0_beacon-chain.md [hunk 9]
```diff
@@ -17,19 +17,28 @@
...
         - [Reward and penalty quotients](#reward-and-penalty-quotients)
         - [Status codes](#status-codes)
-        - [Special record types](#special-record-types)
+        - [Max operations per block](#max-operations-per-block)
         - [Validator registry delta flags](#validator-registry-delta-flags)
         - [Signature domains](#signature-domains)
     - [Data structures](#data-structures)
-        - [Deposits](#deposits)
-            - [`DepositParametersRecord`](#depositparametersrecord)
+        - [Beacon chain operations](#beacon-chain-operations)
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: Not sure if the following categories are very accurate.
1. `AttestationData` is also used in `CasperVotes`.
2. If we put all `*Data` class to here, Why `ProposalSignedData` is still in `ProposalSignedData` section?

File: specs/core/0_beacon-chain.md [hunk 10]
```diff
@@ -226,54 +232,58 @@ Unless otherwise indicated, code appearing in `this style` is to be interpreted
...
-    'body': BeaconBlockBody,
+    # First batch of votes
+    'votes_1': CasperVotes,
+    # Second batch of votes
+    'votes_2': CasperVotes,
 }
 ```
 
-#### `BeaconBlockBody`
+##### `CasperVotes`
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: IMO plural `Votes` is confusing sometimes. Per @djrtwo: `SlashableAttestationData`? Or something like `VoteData`?

File: specs/core/0_beacon-chain.md [hunk 11]
```diff
@@ -311,27 +321,101 @@ Unless otherwise indicated, code appearing in `this style` is to be interpreted
...
 
 ```python
 {
-    # Slot number
+    # Receipt Merkle branch
+    'merkle_branch': '[hash32]',
+    # Merkle tree index
+    'merkle_tree_index': 'uint64',
+    # Deposit data
+    'deposit_data': {
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: Whoops, I didn't notice this before. Maybe move it to a new structure? (In other PRs)

File: specs/core/0_beacon-chain.md [hunk 12]
```diff
@@ -311,27 +321,101 @@ Unless otherwise indicated, code appearing in `this style` is to be interpreted
...
+        'deposit_parameters': DepositParameters,
+        # Value in Gwei
+        'value': 'uint64',
+        # Timestamp from deposit contract
+        'timestamp': 'uint64',
+    },
+}
+```
+
+##### `DepositParameters`
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: Maybe `DepositInput`?

File: specs/core/0_beacon-chain.md (hunk 11 above)
Comment (djrtwo) [reply specs/core/0_beacon-chain.md]: yeah, separate pr

File: specs/core/0_beacon-chain.md (hunk 12 above)
Comment (djrtwo) [reply specs/core/0_beacon-chain.md]: This mirrors the input arg in the pow contract
```python
def deposit(deposit_parameters: bytes[2048])
```

Going to leave as is. We can change both in a separate PR if you want

**Pull request merged**